Include llvm-link into Windows distribution

|  |  |
| --- | --- |
| Bugzilla Link | [51850](https://llvm.org/bz51850) |
| Version | unspecified |
| OS | Windows NT |
| CC | @xgupta |

## Extended Description 
There is no llvm-link.exe inside Windows distribution, for example in this one https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0-rc2/LLVM-13.0.0-rc2-win64.exe.

I don't know how to join two .bc files into one, like I did with llvm-link on Linux:

llvm-link a.bc b.bc -o ab.bc

Maybe you can suggest at least a workaround with those tools that are included.

Also would be nice to have llvm-as.exe inside package so to be able to compile .ll text assembly to .bc bitcode.

Thanks! When to use the mingw-w64-based vs clang-based vs UCRT-based executables?

I forked the most updated fork of this repo https://github.com/vovkos/llvm-package-windows and adjusted the LLVM version number to 15.0.7, hoping that it would still work, but apparently not:
https://github.com/Boscop/llvm-package-windows/actions/runs/4669816368/jobs/8268772718#step:4:236
```
CMake Error at cmake/modules/CMakeLists.txt:1 (include):
  include could not find requested file:

    ExtendPath


CMake Error at cmake/modules/CMakeLists.txt:3 (include):
  include could not find requested file:

    FindPrefixFromConfig


CMake Error at cmake/modules/CMakeLists.txt:116 (find_prefix_from_config):
  Unknown CMake command "find_prefix_from_config".


-- Configuring incomplete, errors occurred!
MSBUILD : error MSB1009: Project file does not exist.
```
Any idea what changed since version 14 / how the script needs to be adapted?

---

Also, when I tried to build it locally on Win 10, it didn't work either:
```
D:\llvm-15.0.7.src\build>set LLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=ON

D:\llvm-15.0.7.src\build>cmake -GNinja -DCMAKE_BUILD_TYPE=Release ..
[...]
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:327 (include):
  include could not find requested file:

    GNUInstallPackageDir


CMake Error at cmake/modules/CheckCompilerVersion.cmake:44 (message):
  Host Visual Studio version should be at least 19.27 because LLVM will soon
  use new C++ features which your toolchain version doesn't support.  Your
  version is 19.24.28316.0.  You can temporarily opt out using
  LLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN, but very soon your toolchain won't be
  supported.
Call Stack (most recent call first):
  cmake/modules/CheckCompilerVersion.cmake:52 (check_compiler_version)
  cmake/config-ix.cmake:15 (include)
  CMakeLists.txt:776 (include)


-- Configuring incomplete, errors occurred!
```
And I'm also wondering why I still got the error that mentions `LLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN` even though I set it to `ON`. (I have VS 2019 and would prefer to build with that if possible.)

--- a/llvm/utils/release/build_llvm_release.bat
+++ b/llvm/utils/release/build_llvm_release.bat
@@ -287,7 +287,16 @@ ninja check-sanitizer || ninja check-sanitizer || ninja check-sanitizer || exit 
 ninja check-clang-tools || ninja check-clang-tools || ninja check-clang-tools || exit /b 1
 ninja check-clangd || ninja check-clangd || ninja check-clangd || exit /b 1
 ninja package || exit /b 1
+
+:: generate tarball with install toolchain only off
+set filename=clang+llvm-%version%-x86_64-pc-windows-msvc
+cmake -GNinja %cmake_flags% %cmake_profile_flags% -DLLVM_INSTALL_TOOLCHAIN_ONLY=OFF ^
+  -DCMAKE_INSTALL_PREFIX=%build_dir%/%filename% ..\llvm-project\llvm || exit /b 1
+ninja install || exit /b 1
+:: check llvm_config is present & returns something
+%build_dir%/%filename%/bin/llvm-config.exe --bindir || exit /b 1
 cd ..
+7z a -ttar -so %filename%.tar %filename% | 7z a -txz -si %filename%.tar.xz
 
 exit /b 0
 ::==============================================================================

